src/dune_rules/foreign.mli

```diff
@@ -159,12 +172,35 @@ module Source : sig
   val make : stubs:Stubs.t -> path:Path.Build.t -> t
 end
 
+(** The [path] to a foreign object file along with the compilation [mode] that
+    specfies whether it should be included into native or bytecode libraries, or
+    into both (the latter is the most common case). *)
+module Object : sig
+  type t =
+    { path : Path.Build.t
+    ; mode : Compilation_mode.t
+    }
+
+  (** Should this file be included into both native and bytecode libraries? *)
+  val both_byte_and_native : t -> bool
+
+  (** Path to the object file in the build directory. *)
+  val build_path : t -> Path.t
+
+  (** Like [build_path] but returns [Some] only if the object should be included
+      into bytecode libraries. *)
+  val build_path_byte : t -> Path.t option
+
+  (** Like [build_path] but returns [Some] only if the object should be included
+      into native libraries. *)
+  val build_path_native : t -> Path.t option
```

Comment:
We've become quite spartan with our public api's, so I'd suggest to trim helper functions that contain minimal logic. For example, I'd get rid of them all in favour of:

```
val obj_file : t -> Mode.t -> Path.Build.t option
```

The rest can be trivially accessed from the record fields.